### PR TITLE
Cody: Improve chat history

### LINKED ISSUE #52544: Cody: VSCode history should list the current chat
Now we have the [➕] button instead of clear conversation, we need to ensure the current conversation is also at the top of the conversation history list. Because at the moment, the [➕] button is more like a "Clear conversation" function vs the ChatGPT style list of conversations (where the current one is highlighted).

#### Steps to reproduce:

1. Click history, and it's empty, but should list the current conversation

### PR DESCRIPTION
**Pull request by deepak2431**
This PR closes #52544,  #52848, #52889 and #52890.

The issues mostly overlapped, so I handled it in this PR.

Changes made:
1. Add history to the history list on each new conversation started.
2. Display the last human message in history instead of the assistant message.
3. Add a delete button to delete individual chats from history.
4. Load the most recent chat, if available, from the history.
5. Fix the loading of context files.
6. Load the history messages to the end of the scroll position.

## Test plan

All E2E tests have passed.
https://www.loom.com/share/3dca98dd0db94c298df38ab79c5e72d1

### CODE REVIEW

**Comment by deepak2431** [comment]
@abeatrix @toolmantim Requesting review for this PR.

**Comment by abeatrix** [comment]
@deepak2431 Thanks for the ping! I have tagged my team for review and we will get to this asap, thank you so much for the PR ❤️

**Comment by deepak2431** [comment]
> @deepak2431 Thanks for the ping! I have tagged my team for review and we will get to this asap, thank you so much for the PR ❤️

Thanks a lot, @abeatrix. I was looking to add more E2E tests for this. Just a question, how to run the extension in browser mode so I can look over the proper selectors to pick and attach events to it? 
Or is there a different way to debug extensions for getting the selectors to attach test events?

**Comment by abeatrix** [comment]
@deepak2431 if you start the e2e test in debug mode and then hit record in the pop up console, it will show you the name of each selector you hover or click on (all thanks to @philipp-spiess 🙏 )

**Comment by deepak2431** [comment]
> all thanks to @philipp-spiess 🙏 )

@abeatrix Got that. This works fantastic this way as it captures most of it during the record.

**Comment by deepak2431** [comment]
> > all thanks to @philipp-spiess 🙏 )
> 
> @abeatrix Thanks, Got that. This works fantastic as it captures most of it during the record.

**Comment by deepak2431** [comment]
@abeatrix Added the tests. It was fun to write tests by recording the events in sequential order to get the selectors.

**Review by abeatrix — CHANGES_REQUESTED** [review]
Left some comments inline. The new behavior makes a lot more sense to me so thank you so much for working on this 🙇  
In addition, the Clear History button is still clickable if you remove the history item one by one. I tried updating it to `disabled={!userHistory || !Object.keys(userHistory).length}` and it fixed it for me :)

**Comment by deepak2431** [comment]
> Left some comments inline. The new behavior makes a lot more sense to me so thank you so much for working on this 🙇 In addition, the Clear History button is still clickable if you remove the history item one by one. I tried updating it to `disabled={!userHistory || !Object.keys(userHistory).length}` and it fixed it for me :)

Thanks a lot for the review @abeatrix, and for thoroughly testing it. I will make all the requested changes here soon.

File: client/cody/webviews/UserHistory.tsx [hunk 1]
```diff
@@ -79,6 +89,18 @@ export const UserHistory: React.FunctionComponent<React.PropsWithChildren<Histor
                                     >
                                         <div className={styles.itemButtonInnerContainer}>
                                             <div className={styles.itemDate}>{new Date(chat[0]).toLocaleString()}</div>
+                                            <div className={styles.itemDelete}>
+                                                <VSCodeButton
+                                                    appearance="icon"
+                                                    type="button"
+                                                    onClick={event => {
```
[the remark this answers is not in the document]
Comment (deepak2431) [reply client/cody/webviews/UserHistory.tsx:96]: Thanks for pointing this out, I can reproduce it here too.

File: client/cody/webviews/UserHistory.module.css [hunk 2]
```diff
@@ -37,6 +38,17 @@
     margin-bottom: .25rem;
 }
 
+.item-delete {
+    position: absolute;
+    right: 0.25rem;
+    top: 0.25rem;
```
[the remark this answers is not in the document]
Comment (deepak2431) [reply client/cody/webviews/UserHistory.module.css:44]: yeah, that would be better to do.

File: client/cody/src/chat/ChatViewProvider.ts [hunk 3]
```diff
@@ -374,6 +376,7 @@ export class ChatViewProvider implements vscode.WebviewViewProvider, vscode.Disp
         this.cancelCompletionCallback = null
         this.sendTranscript()
         void this.saveTranscriptToChatHistory()
+        this.sendChatHistory()
```
[the remark this answers is not in the document]
Comment (deepak2431) [reply client/cody/src/chat/ChatViewProvider.ts:379]: Yeah, it's calling saveChatHistory(), but the method I have used is sendChatHistory(). So, this should send the current chat instantly the response is received for the started chat or not.

File: client/cody/src/chat/ChatViewProvider.ts [hunk 4]
```diff
@@ -623,6 +626,15 @@ export class ChatViewProvider implements vscode.WebviewViewProvider, vscode.Disp
         this.setWebviewView('login')
     }
 
+    /**
+     * Delete history from current chat history and local storage
+     */
+    private async deleteHistory(chatID: string): Promise<void> {
+        delete this.chatHistory[chatID]
+        await this.localStorage.deleteChatHistory(chatID)
```
[the remark this answers is not in the document]
Comment (deepak2431) [reply client/cody/src/chat/ChatViewProvider.ts:634]: I think both might be the same, as while saving the chat history, we are updating the local storage to rewrite its values getting the chat and input history.  We can go with either was I feel. Let me know what you think would be best way!

File: client/cody/src/chat/ChatViewProvider.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (abeatrix) [reply client/cody/src/chat/ChatViewProvider.ts:379]: I'm not sure too but this is working so we can revisit later if it becomes an issue :)

**Review by abeatrix — APPROVED** [review]

File: client/cody/src/chat/ChatViewProvider.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (deepak2431) [reply client/cody/src/chat/ChatViewProvider.ts:634]: Just curious, what issue do you think it can create later?

**Comment by deepak2431** [comment]
> Approved! Once the conflict is resolved, we can merge this 🎉 thank you so much for the great work!

Thanks a lot, @abeatrix, for the complete review. I have resolved the conflicts so all good to be merged now!

**Comment by abeatrix** [comment]
@deepak2431 i tried to merge but your PR is not passing the e2e tests 🤔

**Comment by deepak2431** [comment]
@abeatrix Hmm, not sure, as it's passing all of them here 👀
iScreen Shoter - Code - 230608131158

**Comment by toolmantim** [comment]
@deepak2431 I think https://github.com/sourcegraph/sourcegraph/pull/53224 should fix the build errors. Can you merge main into this branch, so we can see? Thanks!

**Comment by deepak2431** [comment]
> @deepak2431 I think #53224 should fix the build errors. Can you merge main into this branch, so we can see? Thanks!

@toolmantim I have merged the main and updated the integration test, which I saw now was breaking because we can now load the last chat. So, we need to clear the history after each test gets completed.

All the E2E tests are passing locally. 
All integration tests have passed, too, except the Local file finder one.

I hope now the build will succeed as in the logs it shows E2E tests failing, but I'm not sure why!

**Comment by deepak2431** [comment]
@philipp-spiess @abeatrix @toolmantim  Thanks a lot. Finally, I can see that the required failing build is successful in this PR!

**Pull request merged**